src/databricks/sqlalchemy/dialect.py

```diff
@@ -1,52 +1,301 @@
+import os
+
 from databricks import sql
+from databricks import sql as dbsql
+
+import re
+
+from sqlalchemy import types
+
+# we leverage MySQL's implementation of TINYINT and DOUBLE
+from sqlalchemy.types import Integer, BigInteger, SmallInteger, Float, DECIMAL, Boolean
+from sqlalchemy.types import String, DATE, TIMESTAMP
+
+from sqlalchemy import util
+from sqlalchemy import exc
+
+from sqlalchemy.engine import default, interfaces
+from sqlalchemy.sql import compiler
+
 from typing import AnyStr
 
-from sqlalchemy import util, exc, types
-from sqlalchemy.engine import default
 
+# provide a way to debug
+debugbreakpoint = os.getenv("DATABRICKS_DIALECT_DEBUG") or False
 
-class DatabricksDialect(default.DefaultDialect):
 
+class DatabricksIdentifierPreparer(compiler.IdentifierPreparer):
+    # SparkSQL identifier specification:
+    # ref: https://spark.apache.org/docs/latest/sql-ref-identifier.html
+
+    legal_characters = re.compile(r"^[A-Z0-9_]+$", re.I)
+
+    def __init__(self, dialect):
+        super(DatabricksIdentifierPreparer, self).__init__(
+            dialect,
+            initial_quote="`",
+        )
+
+
+# this class provides visitors that emit the dialect-specific keywords for SQLAlchemy's SQL expression parse tree
+class DatabricksTypeCompiler(compiler.GenericTypeCompiler):
+    # ref: https://spark.apache.org/docs/latest/sql-ref-datatypes.html
+
+    def visit_TINYINT(self, type_):
+        return "TINYINT"
+
+    def visit_SMALLINT(self, type_, **kw):
+        return "SMALLINT"
+
+    def visit_INTEGER(self, type_, **kw):
+        return "INT"
+
+    def visit_BIGINT(self, type_, **kw):
+        return "BIGINT"
+
+    def visit_FLOAT(self, type_, **kw):
+        return "FLOAT"
+
+    def visit_DOUBLE(self, type_, **kw):
+        return "DOUBLE"
+
+    def visit_DECIMAL(self, type_, **kw):
+        if type_.precision is None:
+            return "DECIMAL"
+        elif type_.scale is None:
+            return "DECIMAL(%(precision)s)" % {"precision": type_.precision}
+        else:
+            return "DECIMAL(%(precision)s, %(scale)s)" % {
+                "precision": type_.precision,
+                "scale": type_.scale,
+            }
+
+    def visit_NUMERIC(self, type_, **kw):
+        if type_.precision is None:
+            return "DECIMAL"
+        elif type_.scale is None:
+            return "DECIMAL(%(precision)s)" % {"precision": type_.precision}
```

Comment:
do we have test covering the methods of this class?

Reply:
According to what I found in SA's doc (https://github.com/sqlalchemy/sqlalchemy/blob/main/doc/build/changelog/changelog_05.rst), it "is used for encoding the column name when processing the metadata. This usually defaults to utf-8." I went with "None" as for all the other dialects I studied.

Reply:
Good catch. I will revisit integration.py to include both the default and specified scales.

Reply:
Rectified.